internal/luksview/view.go

```diff
@@ -0,0 +1,197 @@
+)
+
+type namedTokenData struct {
+	id    int
+	token NamedToken
+}
+
+// HeaderSource provides a mechanism to obtain a LUKS2 header.
+type HeaderSource interface {
+	ReadHeader() (*luks2.HeaderInfo, error)
+}
+
+type defaultHeaderSource struct {
+	devicePath string
+	lockMode   luks2.LockMode
+}
+
+func (s *defaultHeaderSource) ReadHeader() (*luks2.HeaderInfo, error) {
+	return luks2.ReadHeader(s.devicePath, s.lockMode)
+}
+
+// View provides a read-only view of a LUKS2 header in a way that is useful
+// to secboot.
+type View struct {
+	source HeaderSource
+
+	hdr *luks2.HeaderInfo
+
+	// namedTokens contains a map of keyslot name to valid named tokens.
+	namedTokens map[string]namedTokenData
+}
+
+// NewView creates a new View from the LUKS2 container at the specified
+// path, using the specified locking mode.
+func NewView(devicePath string, lockMode luks2.LockMode) (*View, error) {
+	view := &View{source: &defaultHeaderSource{
+		devicePath: devicePath,
+		lockMode:   lockMode}}
+
+	if err := view.Refresh(); err != nil {
+		return nil, err
+	}
+
+	return view, nil
+}
+
+// NewViewFromCustomHeaderSource creates a new View from the LUKS2 header
+// obtained from the supplied source. This is useful for mocking the header
+// in unit tests.
+func NewViewFromCustomHeaderSource(source HeaderSource) (*View, error) {
+	view := &View{source: source}
+
+	if err := view.Refresh(); err != nil {
+		return nil, err
+	}
+
+	return view, nil
+}
+
+// Refresh updates this view from the source container.
+func (v *View) Refresh() error {
+	hdr, err := v.source.ReadHeader()
+	if err != nil {
+		return err
+	}
+
+	v.hdr = hdr
+	v.namedTokens = make(map[string]namedTokenData)
+
+	for id, token := range hdr.Metadata.Tokens {
+		named, ok := token.(NamedToken)
+		if !ok {
+			continue
+		}
+
+		if _, exists := v.namedTokens[named.Name()]; exists {
+			return fmt.Errorf("multiple tokens with the same name (%s)", named.Name())
+		}
+
+		v.namedTokens[named.Name()] = namedTokenData{id: id, token: named}
+	}
+
+	return nil
+}
+
+// ListNames returns a sorted list of all of the keyslot names from this view.
+func (v *View) ListNames() (names []string) {
+	for name, data := range v.namedTokens {
+		if _, orphaned := data.token.(*orphanedToken); orphaned {
+			continue
+		}
+		names = append(names, name)
+	}
+	sort.Strings(names)
+	return names
+}
+
+// TokenByName returns the token and its ID for the keyslot with the supplied name.
+func (v *View) TokenByName(name string) (token NamedToken, id int, exists bool) {
+	data, exists := v.namedTokens[name]
+	if !exists {
+		return nil, 0, false
+	}
+	if _, orphaned := data.token.(*orphanedToken); orphaned {
+		return nil, 0, false
+	}
+	return data.token, data.id, true
+}
+
+// KeyDataTokensByPriority returns all of the key data tokens in order of priority,
+// from highest to lowest. Tokens with the same priority are returned in the order in
+// which their names are sorted. This omits any with a priority of 0.
```

Comment:
the code and the comment don't seem to match. The code has priority < 0 (not <= 0)

Reply:
I've updated this.